pkg/ccl/backupccl/backup_test.go

```diff
@@ -1411,177 +1411,85 @@ into_db='restoredb', %s)`, encryptionOption), backupLoc1)
-			// backup issues. DistSender will still split those requests to different
-			// ranges on the same node. Each range will write a file, so the number of
-			// SST files this backup will write is `backupRestoreDefaultRanges` .
-			totalExpectedResponses = totalExpectedBackupRequests
-		} else if strings.Contains(query, "RESTORE") {
-			// We expect restore to process each file in the backup individually.
-			// SST files are written per-range in the backup. So we expect the
-			// restore to process #(ranges) that made up the original table.
-			totalExpectedResponses = backupRestoreDefaultRanges
-		} else {
-			t.Fatal("expected query to be either a backup or restore")
-		}
-		jobDone := make(chan error)
-		allowResponse = make(chan struct{}, totalExpectedResponses)
-
-		go func() {
-			_, err := conn.Exec(query, localFoo)
-			jobDone <- err
-		}()
-
-		// Allow half the total expected responses to proceed.
-		for i := 0; i < totalExpectedResponses/2; i++ {
-			allowResponse <- struct{}{}
-		}
-
-		// Due to ExportRequest pagination, in the case of backup, we want to wait
-		// until an entire span has been exported before checking job progress.
-		if strings.Contains(query, "BACKUP") {
-			exportSpanCompleteCh <- struct{}{}
-			close(exportSpanCompleteCh)
+	defer cleanupFn()
+	var jobID jobspb.JobID
+	sqlDB.QueryRow(t, `BACKUP INTO 'userfile:///foo' WITH detached`).Scan(&jobID)
+	jobutils.WaitForJobToRun(t, sqlDB, jobID)
+
+	// Allow a few requests to go through.
+	blockCh <- struct{}{}
+	blockCh <- struct{}{}
+	blockCh <- struct{}{}
+
+	defer func() {
+		close(blockCh)
+		jobutils.WaitForJobToSucceed(t, sqlDB, jobID)
+		var finalProgress float32
+		sqlDB.QueryRow(t,
+			`SELECT fraction_completed FROM crdb_internal.jobs WHERE job_id = $1`,
+			jobID,
+		).Scan(&finalProgress)
+		require.Equal(t, float32(1), finalProgress)
+	}()
+	var fractionCompleted float32
+	testutils.SucceedsSoon(t, func() error {
+		sqlDB.QueryRow(t,
+			`SELECT fraction_completed FROM crdb_internal.jobs WHERE job_id = $1`,
+			jobID,
+		).Scan(&fractionCompleted)
+		if fractionCompleted < 0.01 || fractionCompleted > 0.99 {
+			return errors.Newf(
+				"expected progress to be in range [0.01, 0.99] but got %f",
+				fractionCompleted,
+			)
 		}
-		err := retry.ForDuration(testutils.DefaultSucceedsSoonDuration, func() error {
-			return check(ctx, inProgressState{
-				DB:            conn,
-				backupTableID: backupTableID,
-				dir:           dir,
-				name:          "foo",
-			})
-		})
-
-		// Close the channel to allow all remaining responses to proceed. We do this
-		// even if the above retry.ForDuration failed, otherwise the test will hang
-		// forever.
-		close(allowResponse)
+		return nil
+	})
 
-		if err := <-jobDone; err != nil {
-			t.Fatalf("%q: %+v", query, err)
-		}
+	// Allow some more requests to go through.
+	blockCh <- struct{}{}
+	blockCh <- struct{}{}
+	blockCh <- struct{}{}
 
-		if err != nil {
-			t.Fatal(err)
+	var newFractionCompleted float32
```

Comment:
[nit] If this is only used in the scope of the succeeds soon func, I think it would be clearer to define it there.

I could see adding a small helper function `fractionCompletedForJob(testing.T, jobspb.JobID) float32`

Reply:
moved the var inside the closure